waku/v2/node/peer_manager/peer_manager.nim

```diff
@@ -54,45 +64,65 @@ proc insertOrReplace(ps: PeerStorage,
 proc dialPeer(pm: PeerManager, peerId: PeerID,
               addrs: seq[MultiAddress], proto: string,
               dialTimeout = DefaultDialTimeout,
-              source = "api"
+              source = "api",
+              respectBackoff = false,
               ): Future[Option[Connection]] {.async.} =
 
   # Do not attempt to dial self
   if peerId == pm.switch.peerInfo.peerId:
     return none(Connection)
 
-  info "Dialing peer from manager", wireAddr = addrs, peerId = peerId
+  let failedAttempts = pm.peerStore[NumberFailedConnBook][peerId]
+
+  # Respect a backoff period for peers that have failed to connect
+  if respectBackoff and not pm.peerStore.canBeConnected(peerId, pm.initialBackoffInSec):
+    debug "Peer recently failed to connect, skipping dial",
+          peerId = peerid,
+          failedAttempts = failedAttempts,
+          lastFailed = pm.peerStore[LastFailedConnBook][peerId]
+    return none(Connection)
+
+  debug "Dialing peer from manager", wireAddr = addrs, peerId = peerId, failedAttempts=failedAttempts
 
   # Dial Peer
+  var deadline = sleepAsync(dialTimeout)
   let dialFut = pm.switch.dial(peerId, addrs, proto)
 
+  var reasonFailed = ""
   try:
-    # Attempt to dial remote peer
-    if (await dialFut.withTimeout(DefaultDialTimeout)):
+    await dialFut or deadline
+    if dialFut.finished():
+      if not deadline.finished():
+        deadline.cancel()
       waku_peers_dials.inc(labelValues = ["successful"])
       # TODO: This will be populated from the peerstore info when ready
       waku_node_conns_initiated.inc(labelValues = [source])
       return some(dialFut.read())
     else:
-      # TODO: any redial attempts?
-      debug "Dialing remote peer timed out"
-      waku_peers_dials.inc(labelValues = ["timeout"])
-
-      pm.peerStore[ConnectionBook][peerId] = CannotConnect
-      if not pm.storage.isNil:
-        pm.storage.insertOrReplace(peerId, pm.peerStore.get(peerId), CannotConnect)
+      reasonFailed = "timeout"
+      await cancelAndWait(dialFut)
+  except CatchableError as exc:
+    reasonFailed = "failed"
+
+  # Dial failed
+  pm.peerStore[NumberFailedConnBook][peerId] = pm.peerStore[NumberFailedConnBook][peerId] + 1
+  pm.peerStore[LastFailedConnBook][peerId] = Moment.init(getTime().toUnix, Second)
+  pm.peerStore[ConnectionBook][peerId] = CannotConnect
+
+  debug "Dialing remote failed", peerId = peerId, reason = reasonFailed, failedAttempts=failedAttempts
+  waku_peers_dials.inc(labelValues = [reasonFailed])
+
+  # If failed too many times, remove peer from peer store
+  if respectBackoff and pm.peerStore[NumberFailedConnBook][peerId] >= pm.maxFailedAttempts:
```

Comment:
This seems a bit weird to me - the fact that me dialing some peer could result in it being removed from the peer store (and doing it again would result in it being added again, presumably?). Of course, we know that for the _application_ of "continuously attempt to connect to all available `relay` peers" this would make sense, i.e. to eventually stop attempting to dial peers that continue to fail. But this highlights my concern that this logic should not be part of the dialer, even if behind a `respectBackoff` flag.

Reply:
fixed